// resuite/src/Arrow.js
import React, { PropTypes } from 'react';
import { Svg } from '.';

const rotation = {
  up: '180deg',
  down: '0deg',
  left: '90deg',
  right: '-90deg'
};

const Arrow = (props) => {
  const style = {
    transformOrigin: 'center',
    transform: `rotate(${rotation[props.direction]})`,
    display: 'inline-block',
    verticalAlign: 'middle',
    height: props.height,
    width: props.width
  };
  return (
    <div className={'Arrow'} style={style}>
      <Svg width={props.width} height={props.height} viewBox={'0 0 50 50'}>
        <circle cx='25' cy='25' r='24' fill={props.fill} />
        <path 
          d='M10.5 19.5L25.2 34.2'
          stroke={props.stroke} 
          strokeWidth={props.strokeWidth}
          strokeLinecap={props.rounded ? 'round' : 'square'} />
        <path 
          d='M39.9 19.5L25.2 34.2' 
          stroke={props.stroke} 
          strokeWidth={props.strokeWidth}
          strokeLinecap={props.rounded ? 'round' : 'square'} />
      </Svg>
    </div>
  );
};

Arrow.propTypes = {
  direction:   PropTypes.oneOf(['up', 'down', 'left', 'right']),
  stroke:      PropTypes.string,
  fill:        PropTypes.string,
  strokeWidth: PropTypes.number,
  width:       PropTypes.number,
  height:      PropTypes.number,
  rounded:     PropTypes.bool
};

Arrow.defaultProps = {
  direction: 'down',
  stroke: 'currentColor',
  fill: 'transparent',
  strokeWidth: 5,
  width: 50,
  height: 50,
  rounded: false
};

export default Arrow;
