// resuite/src/Image.js
import React, { PropTypes, createElement } from 'react';
import Preload from 'repreload';
import vudu from 'vudu';

const Image = (props) => {
  const styles = vudu({
    wrapper: {
      position: 'relative',
      ':before': {
        display: 'block',
        content: '""',
        width: props.width,
        paddingTop: '100%'
      }
    },
    content: {
      backgroundSize: 'cover',
      backgroundPosition: '50% 50%',
      backgroundRepeat: 'no-repeat',
      backgroundImage: `url(${props.src})`,
      position: 'absolute',
      bottom: 0,
      right: 0,
      left: 0,
      top: 0
    }
  });
  return (
    <Preload {...props}>
      <div className={styles.wrapper}>
        <div className={styles.content}></div>
      </div>
    </Preload>
  );
};

Image.propTypes = {
  src: PropTypes.string.isRequired,
  loadingBackground: PropTypes.string,
  loadingIndicator: PropTypes.element,
  errorNotice: PropTypes.element,
  width: PropTypes.string
};

Image.defaultProps = {
  src: '',
  loadingBackground: '#eeeeee',
  loadingIndicator: <span>Loading...</span>,
  errorNotice: <span>Oops, couldn’t load image</span>,
  width: '100%'
};

export default Image;
